dbms/src/Common/QueryProfiler.h

```diff
@@ -0,0 +1,158 @@
+{
+    /// Normally query_id is a UUID (string with a fixed length) but user can provide custom query_id.
+    /// Thus upper bound on query_id length should be introduced to avoid buffer overflow in signal handler.
+    constexpr size_t QUERY_ID_MAX_LEN = 1024;
+
+    void writeTraceInfo(TimerType timer_type, int /* sig */, siginfo_t * /* info */, void * context)
+    {
+        constexpr size_t buf_size = sizeof(char) + // TraceCollector stop flag
+                                    8 * sizeof(char) + // maximum VarUInt length for string size
+                                    QUERY_ID_MAX_LEN * sizeof(char) + // maximum query_id length
+                                    sizeof(StackTrace) + // collected stack trace
+                                    sizeof(TimerType); // timer type
+        char buffer[buf_size];
+        DB::WriteBufferFromFileDescriptor out(trace_pipe.fds_rw[1], buf_size, buffer);
+
+        StringRef query_id = CurrentThread::getQueryId();
+        query_id.size = std::min(query_id.size, QUERY_ID_MAX_LEN);
+
+        const auto signal_context = *reinterpret_cast<ucontext_t *>(context);
+        const StackTrace stack_trace(signal_context);
+
+        DB::writeChar(false, out);
+        DB::writeStringBinary(query_id, out);
+        DB::writePODBinary(stack_trace, out);
+        DB::writePODBinary(timer_type, out);
+        out.next();
+    }
+
+    const UInt32 TIMER_PRECISION = 1e9;
+}
+
+
+/**
+  * Query profiler implementation for selected thread.
+  *
+  * This class installs timer and signal handler on creation to:
+  *  1. periodically pause given thread
+  *  2. collect thread's current stack trace
+  *  3. write collected stack trace to trace_pipe for TraceCollector
+  *
+  * Desctructor tries to unset timer and restore previous signal handler.
+  * Note that signal handler implementation is defined by template parameter. See QueryProfilerReal and QueryProfilerCpu.
+  */
+template <typename ProfilerImpl>
+class QueryProfilerBase
+{
+public:
+    QueryProfilerBase(const Int32 thread_id, const int clock_type, const UInt32 period, const int pause_signal = SIGALRM)
+        : log(&Logger::get("QueryProfiler"))
+        , pause_signal(pause_signal)
+    {
+        struct sigaction sa{};
+        sa.sa_sigaction = ProfilerImpl::signalHandler;
+        sa.sa_flags = SA_SIGINFO | SA_RESTART;
+
+        if (sigemptyset(&sa.sa_mask))
+            throw Poco::Exception("Failed to clean signal mask for query profiler");
+
+        if (sigaddset(&sa.sa_mask, pause_signal))
+            throw Poco::Exception("Failed to add signal to mask for query profiler");
+
+        if (sigaction(pause_signal, &sa, previous_handler))
+            throw Poco::Exception("Failed to setup signal handler for query profiler");
+
+        struct sigevent sev;
+        sev.sigev_notify = SIGEV_THREAD_ID;
+        sev.sigev_signo = pause_signal;
+        sev._sigev_un._tid = thread_id;
+        if (timer_create(clock_type, &sev, &timer_id))
```

Comment:
Exception safety when signal handler is installed but we cannot set timer:
-- destructor won't be called and we have to try to rollback signal handler.